### DISCUSSION TITLE
Dagster Cloud Serverless sensors timing out when running in parallel

### DISCUSSION THREAD

**Discussion by misa-nogly-aleksic** [dagster-io/dagster#19188] [Q&A]
Has anyone experienced this before and has any tips?

I'm using `dagster v1.5.14` and `dagster-cloud v1.5.14`.

### Observation

I have 6 sensors running in one code location, each one querying S3 for new files (since a cursor). For this we just use `from dagster_aws.s3.sensor import get_s3_keys`.

**When I enable all of them**, they start timing out. I get the following error:
```
dagster._core.errors.DagsterUserCodeUnreachableError: dagster._core.errors.DagsterUserCodeUnreachableError: 
The sensor tick timed out due to taking longer than 60 seconds to execute the sensor function. 
One way to avoid this error is to break up the sensor work into chunks, using cursors to let subsequent 
sensor calls pick up where the previous call left off.

Stack Trace:
  File "/dagster-cloud/dagster_cloud/agent/dagster_cloud_agent.py", line 843, in _process_api_request
    api_result = self._handle_api_request(
  File "/dagster-cloud/dagster_cloud/agent/dagster_cloud_agent.py", line 703, in _handle_api_request
    serialized_sensor_data_or_error = client.external_sensor_execution(
  File "/dagster/dagster/_grpc/client.py", line 466, in external_sensor_execution
    return self._query(
  File "/dagster/dagster/_grpc/client.py", line 173, in _query
    self._raise_grpc_exception(
  File "/dagster/dagster/_grpc/client.py", line 151, in _raise_grpc_exception
    raise DagsterUserCodeUnreachableError(

The above exception was caused by the following exception:
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.DEADLINE_EXCEEDED
	details = "Deadline Exceeded"
	debug_error_string = "UNKNOWN:Error received from peer  {created_time:"2024-01-12T15:13:48.280017327+00:00", grpc_status:4, grpc_message:"Deadline Exceeded"}"
>

```

**When I enable one at a time** they work just fine and take 10-40s to evaluate.

### Consulting the Docs

When I consult the docs here:
https://docs.dagster.io/concepts/partitions-schedules-sensors/sensors#sensor-optimizations-using-cursors

> If a sensor evaluation function takes more than 60 seconds to return its results, the sensor evaluation **will time out and the Dagster daemon will move on to the next sensor** without submitting any runs. This 60 second timeout only applies to the time it takes to run the sensor function, not to the execution time of the runs submitted by the sensor. To avoid timeouts, slower sensors can break up their work into chunks, using [cursors](https://docs.dagster.io/concepts/partitions-schedules-sensors/sensors#sensor-optimizations-using-cursors) to let subsequent sensor calls pick up where the previous call left off.

I get the impression that sensors **are evaluated in series.**

But this contradicts my observations. 

### Resolution / Workaround

I'm working around this by setting the `default_sensor_timeout` to 5 minutes in my deployment configuration. Now my sensors don't time out.

But this doesn't feel right. We don't have very many files in this S3 bucket (<10k), and querying S3 locally takes about 5 seconds, not 1 minute.

**Comment by ytoast** [comment]
Did you manage to resolve this issue? I have increase my default_sensor_timeout to 10minutes, and it still times out. Would increasing the sensor intervals help? (maybe from 30s to 300s)

I do have lots of sensors running at the same time, and I wonder if there is an upper limit to the number of sensors that we can run in each code location. What might be the best way to debug this and to find out which part of the sensor is causing the delay?

**Comment by garethbrickman** [comment]
There is an extensive guide for debugging sensor timeouts in https://github.com/dagster-io/dagster/discussions/20021
